patches/common/swiftshader/fixed_all_oob_accesses_in_vertexprogram_and_pixelprogram.patch

```diff
@@ -0,0 +1,508 @@
+ 		// FIXME: Use enableLeave in other control-flow constructs
+diff --git a/src/Shader/PixelProgram.hpp b/src/Shader/PixelProgram.hpp
+index 240938dd15820601ce2bf5e4ff6eb242ddd196e7..4ed3eef7545e27b56ceffa5a7cf33ebe9b7c287f 100644
+--- a/src/Shader/PixelProgram.hpp
++++ b/src/Shader/PixelProgram.hpp
+@@ -27,7 +27,7 @@ namespace sw
+ 			PixelRoutine(state, shader), r(shader->indirectAddressableTemporaries),
+ 			loopDepth(-1), ifDepth(0), loopRepDepth(0), currentLabel(-1), whileTest(false)
+ 		{
+-			for(int i = 0; i < 2048; ++i)
++			for(int i = 0; i < MAX_SHADER_CALL_SITES; ++i)
+ 			{
+ 				labelBlock[i] = 0;
+ 			}
+@@ -67,17 +67,17 @@ namespace sw
+ 
+ 		// DX9 specific variables
+ 		Vector4f p0;
+-		Array<Int, 4> aL;
+-		Array<Int, 4> increment;
+-		Array<Int, 4> iteration;
++		Array<Int, MAX_SHADER_NESTED_LOOPS> aL;
++		Array<Int, MAX_SHADER_NESTED_LOOPS> increment;
++		Array<Int, MAX_SHADER_NESTED_LOOPS> iteration;
+ 
+ 		Int loopDepth;    // FIXME: Add support for switch
+ 		Int stackIndex;   // FIXME: Inc/decrement callStack
+-		Array<UInt, 16> callStack;
++		Array<UInt, MAX_SHADER_CALL_STACK_SIZE> callStack;
+ 
+ 		// Per pixel based on conditions reached
+ 		Int enableIndex;
+-		Array<Int4, 1 + 24> enableStack;
++		Array<Int4, MAX_SHADER_ENABLE_STACK_SIZE> enableStack;
+ 		Int4 enableBreak;
+ 		Int4 enableContinue;
+ 		Int4 enableLeave;
+@@ -152,18 +152,18 @@ namespace sw
+ 		void RET();
+ 		void LEAVE();
+ 
+-		int ifDepth;
+-		int loopRepDepth;
+-		int currentLabel;
++		BoundedIndex<MAX_SHADER_NESTED_IFS> ifDepth = 0;
++		BoundedIndex<MAX_SHADER_NESTED_LOOPS> loopRepDepth = 0;
++		BoundedIndex<MAX_SHADER_CALL_SITES> currentLabel = -1;
+ 		bool whileTest;
+ 
+-		BasicBlock *ifFalseBlock[24 + 24];
+-		BasicBlock *loopRepTestBlock[4];
+-		BasicBlock *loopRepEndBlock[4];
+-		BasicBlock *labelBlock[2048];
+-		std::vector<BasicBlock*> callRetBlock[2048];
++		BasicBlock *ifFalseBlock[MAX_SHADER_NESTED_IFS];
++		BasicBlock *loopRepTestBlock[MAX_SHADER_NESTED_LOOPS];
++		BasicBlock *loopRepEndBlock[MAX_SHADER_NESTED_LOOPS];
++		BasicBlock *labelBlock[MAX_SHADER_CALL_SITES];
++		std::vector<BasicBlock*> callRetBlock[MAX_SHADER_CALL_SITES];
+ 		BasicBlock *returnBlock;
+-		bool isConditionalIf[24 + 24];
++		bool isConditionalIf[MAX_SHADER_NESTED_IFS];
+ 	};
+ }
+ 
+diff --git a/src/Shader/Shader.cpp b/src/Shader/Shader.cpp
+index 36192c93c7473e7c4bc140843ad2c16ca11d3788..0e06703cae08a5e955684d7dc3f03069c3789f4f 100644
+--- a/src/Shader/Shader.cpp
++++ b/src/Shader/Shader.cpp
+@@ -1877,13 +1877,13 @@ namespace sw
+ 	// This is used to know what basic block to return to.
+ 	void Shader::analyzeCallSites()
+ 	{
+-		int callSiteIndex[2048] = {0};
++		int callSiteIndex[MAX_SHADER_CALL_SITES] = {0};
+ 
+ 		for(auto &inst : instruction)
+ 		{
+ 			if(inst->opcode == OPCODE_CALL || inst->opcode == OPCODE_CALLNZ)
+ 			{
+-				int label = inst->dst.label;
++				int label = sw::min(inst->dst.label, unsigned int(MAX_SHADER_CALL_SITES));
```

Comment:
This line does not compile:
```
../../third_party/swiftshader/src/Shader/Shader.cpp:1886:51: error: expected '(' for function-style cast or type construction
                                int label = sw::min(inst->dst.label, unsigned int(MAX_SHADER_CALL_SITES));
                                                                     ~~~~~~~~ ^
1 error generated.
```

Replacing `unsigned int` with `unsigned` should be able to fix it.